src/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.cpp

```diff
@@ -315,27 +355,58 @@ Propagator Propagator::FromEnvironment(const NumericalSolver& aNumericalSolver,
     };
 }
 
-void Propagator::registerDynamicsContext(const Shared<Dynamics>& aDynamicsSPtr)
+void Propagator::validateDynamicsSet() const
 {
-    // Store read coordinate subsets information
-    Array<Pair<Index, Size>> readInfo = Array<Pair<Index, Size>>::Empty();
-    readInfo.reserve(aDynamicsSPtr->getReadCoordinateSubsets().getSize());
-    for (const Shared<const CoordinateSubset>& subset : aDynamicsSPtr->getReadCoordinateSubsets())
+    Map<std::type_index, Array<Shared<Dynamics>>> dynamicsMap;
+
+    for (const Dynamics::Context& dynamicsContext : dynamicsContexts_)
     {
-        const Pair<Index, Size> indexAndSize = {this->coordinatesBrokerSPtr_->addSubset(subset), subset->getSize()};
-        readInfo.add(indexAndSize);
+        (dynamicsMap[typeid(*dynamicsContext.dynamics)]).add(dynamicsContext.dynamics);
     }
 
-    // Store write coordinate subsets information
-    Array<Pair<Index, Size>> writeInfo = Array<Pair<Index, Size>>::Empty();
-    writeInfo.reserve(aDynamicsSPtr->getWriteCoordinateSubsets().getSize());
-    for (const Shared<const CoordinateSubset>& subset : aDynamicsSPtr->getWriteCoordinateSubsets())
+    // Check for mandatory dynamics
+    try
     {
-        const Pair<Index, Size> indexAndSize = {this->coordinatesBrokerSPtr_->addSubset(subset), subset->getSize()};
-        writeInfo.add(indexAndSize);
+        if (dynamicsMap.at(std::type_index(typeid(CentralBodyGravity))).getSize() != 1)
+        {
+            throw ostk::core::error::RuntimeError("Propagator needs exactly one Central Body Gravity Dynamics.");
+        }
+        if (dynamicsMap.at(std::type_index(typeid(PositionDerivative))).getSize() != 1)
+        {
+            throw ostk::core::error::RuntimeError("Propagator needs exactly one Position Derivative Dynamics.");
+        }
+    }
+    catch (const std::out_of_range& e)
```

Comment:
Nit: you can use map's `.find()` to check if a key exists, this way you avoid nesting/repeating error messages:

std::map<int, std::string> myMap;

```cpp
    // Insert some key-value pairs into the map
    myMap[1] = "One";
    myMap[2] = "Two";
    myMap[3] = "Three";

    // Check if the map contains a key
    int keyToFind = 2;
    auto it = myMap.find(keyToFind);
    if (it != myMap.end()) {
        std::cout << "Key " << keyToFind << " found in the map. Value: " << it->second << std::endl;
    } else {
        std::cout << "Key " << keyToFind << " not found in the map." << std::endl;
    }
```

Reply:
Lol that's what I did at first, and then I undid it and replaced it with `try catch` cause I thought it made the code more readable 😢 😭 .

Reply:
But good idea, i'll use the iterator itself to check the value like you did in your example